[Fix apache/incubator-kie-issues#1904] Add audit log messages when persiting/restoring headers

gabriel-farache opened the pull request:
Fix https://github.com/apache/incubator-kie-issues/issues/1904

fjtirado commented on jbpm/process-serialization-protobuf/src/main/java/org/jbpm/flow/serialization/impl/ProtobufProcessInstanceReader.java:
```diff
@@ -177,6 +177,9 @@ private RuleFlowProcessInstance buildWorkflow(KogitoProcessInstanceProtobuf.Proc
 
         if (processInstanceProtobuf.getHeadersList() != null) {
             processInstance.setHeaders(processInstanceProtobuf.getHeadersList().stream().collect(Collectors.toMap(HeaderEntry::getKey, HeaderEntry::getValueList)));
+            processInstance.getHeaders().forEach((key, values) -> {
+                LOGGER.info("Header {} restored for process instance {}", key, processInstance.getId());
+            });
```
```suggestion
                LOGGER.debug("Headers {} restored for process instance {}", processInstance.getHeaders().keySet(), processInstance.getId()));
           }
```

fjtirado reviewed (commented):
I think is better to not mix filter logic with debugging logic. 
Also, log level should be debug, not info.

gabriel-farache commented on jbpm/process-serialization-protobuf/src/main/java/org/jbpm/flow/serialization/impl/ProtobufProcessInstanceWriter.java:
```diff
@@ -164,6 +161,21 @@ public void writeProcessInstance(WorkflowProcessInstanceImpl workFlow, OutputStr
         }
     }
 
+    private Stream<Entry<String, List<String>>> getHeadersStream(WorkflowProcessInstanceImpl workFlow, HeadersPersistentConfig headersConfig) {
+        Stream<Entry<String, List<String>>> stream = workFlow.getHeaders().entrySet().stream();
+        if (headersConfig.excluded() != null && !headersConfig.excluded().isEmpty()) {
+            stream = stream.filter(e -> {
+                if (headersConfig.excluded().contains(e.getKey())) {
+                    LOGGER.info("Excluding header {} from process instance {} from persistence", e.getKey(), workFlow.getId());
+                    return false;
+                }
+                LOGGER.info("Adding header {} from process instance {} for persistence", e.getKey(), workFlow.getId());
+                return true;
+             });
+        }
+        return stream;
+    }
```
just to understand: it's your coding practice to not limit the size of the methods? I tend to try to limit it to 40 and to extract sub-method as much as I can, hence this new method.
I am fine not doing it though, as long as it matches your practices :)

gabriel-farache commented on jbpm/process-serialization-protobuf/src/main/java/org/jbpm/flow/serialization/impl/ProtobufProcessInstanceWriter.java:
```diff
@@ -143,11 +143,8 @@ public void writeProcessInstance(WorkflowProcessInstanceImpl workFlow, OutputStr
-                stream = stream.filter(e -> !headersConfig.excluded().contains(e.getKey()));
-            }
-            instance.addAllHeaders(stream.map(e -> HeaderEntry.newBuilder().setKey(e.getKey()).addAllValue(e.getValue()).build()).collect(Collectors.toList()));
+            Stream<Entry<String, List<String>>> headersStream = getHeadersStream(workFlow, headersConfig);
+            instance.addAllHeaders(headersStream.map(e -> HeaderEntry.newBuilder().setKey(e.getKey()).addAllValue(e.getValue()).build()).collect(Collectors.toList()));
```
and instead of printing the excluded headers when found, you prefer logging first all the headers in the process instance and then the persisted ones? I find it more difficult when debugging an issue to do the non-intersect than having what was excluded and what is really persisted

ricardozanini replied:
I think he added `LOGGER.isDebugEnabled` condition because the collector's operation will be computed despite the `LOGGER.debug` assertion.

fjtirado replied:
I dont think is acceptable to use a filter stream operation (which is expected to just return false/true) to also execute debug statement. 
Also, there was a logic issue, the filter was only executed when excludedheaders are specified, so, in an scenario where there is not excludedheaders you wont see which headers are stored at all.

gabriel-farache commented:
@fjtirado could you review it again?

ricardozanini commented:
@gabriel-farache, @fjtirado is on PTO until May 30th.

ricardozanini commented:
Adding @gmunozfe to the loop.

gmunozfe reviewed (changes requested):
@gabriel-farache hi, could you check this formatting error and try to fix the pipelines to be green?

` Error:  Failed to execute goal net.revelc.code.formatter:formatter-maven-plugin:2.13.0:validate (default) on project process-serialization-protobuf: File '/home/runner/work/incubator-kie-kogito-runtimes/incubator-kie-kogito-runtimes/apache_incubator-kie-kogito-runtimes/jbpm/process-serialization-protobuf/src/main/java/org/jbpm/flow/serialization/impl/ProtobufProcessInstanceReader.java' has not been previously formatted.  Please format file and commit before running validation!`

gabriel-farache commented:
> jbpm/process-serialization-protobuf/src/main/java/org/jbpm/flow/serialization/impl/ProtobufProcessInstanceReader.java

@gmunozfe  When I ran
```
mvn -f jbpm/process-serialization-protobuf net.revelc.code.formatter:formatter-maven-plugin:2.13.0:format
```
I got changes only on 
```
        modified:   jbpm/process-serialization-protobuf/src/main/java/org/jbpm/flow/serialization/protobuf/KogitoNodeInstanceContentsProtobuf.java
        modified:   jbpm/process-serialization-protobuf/src/main/java/org/jbpm/flow/serialization/protobuf/KogitoProcessInstanceProtobuf.java
        modified:   jbpm/process-serialization-protobuf/src/main/java/org/jbpm/flow/serialization/protobuf/KogitoTypesProtobuf.java
        modified:   jbpm/process-serialization-protobuf/src/main/java/org/jbpm/flow/serialization/protobuf/KogitoWorkItemsProtobuf.java
```
Nothing for `pache_incubator-kie-kogito-runtimes/jbpm/process-serialization-protobuf/src/main/java/org/jbpm/flow/serialization/impl/ProtobufProcessInstanceReader.java`, did I use the right command?
